protos/controllers.proto

```diff
@@ -0,0 +1,15 @@
+syntax = "proto3";
+
+package dcs;
+
+message SetAlarmStateRequest {
+	enum AlarmState {
+		AUTO = 0;
+		GREEN = 1;
+		RED = 2;
+	}
+
+	optional string group_name = 1;
+	optional string unit_name = 2;
+	AlarmState alarm_state = 3;
```

Comment:
> I am unable to actually run this as I am getting an exception building. could you have a look and see if I have missed anything obvious? I have stared at the code for a while and cannot see it:
> 
> ```
> Caused by:
>   process didn't exit successfully: `C:\Users\jeff\source\repos\rust-server\target\debug\build\dcs-grpc-server-7638a43898b30456\build-script-build` (exit code: 101)
>   --- stderr
>   thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', C:\Users\jeff\.cargo\registry\src\github.com-1ecc6299db9ec823\prost-build-0.7.0\src\code_generator.rs:259:47
>   stack backtrace:
>      0: std::panicking::begin_panic_handler
>                at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633\/library\std\src\panicking.rs:515
>      1: core::panicking::panic_fmt
>                at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633\/library\core\src\panicking.rs:92
>      2: core::panicking::panic
>                at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633\/library\core\src\panicking.rs:50
>      3: enum$<core::option::Option<alloc::vec::Vec<tuple<prost_types::FieldDescriptorProto, usize>, alloc::alloc::Global>>, 1, 18446744073709551615, Some>::unwrap<alloc::vec::Vec<tuple<prost_types::FieldDescriptorProto, usize>, alloc::alloc::Global>>
>                at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633\library\core\src\option.rs:388
>      4: prost_build::code_generator::CodeGenerator::append_message
>                at C:\Users\jeff\.cargo\registry\src\github.com-1ecc6299db9ec823\prost-build-0.7.0\src\code_generator.rs:259
>      5: prost_build::code_generator::CodeGenerator::generate
>                at C:\Users\jeff\.cargo\registry\src\github.com-1ecc6299db9ec823\prost-build-0.7.0\src\code_generator.rs:102
>      6: prost_build::Config::generate
>                at C:\Users\jeff\.cargo\registry\src\github.com-1ecc6299db9ec823\prost-build-0.7.0\src\lib.rs:766
>      7: prost_build::Config::compile_protos<str>
>                at C:\Users\jeff\.cargo\registry\src\github.com-1ecc6299db9ec823\prost-build-0.7.0\src\lib.rs:725
>      8: tonic_build::prost::Builder::compile_with_config<str>
>                at C:\Users\jeff\.cargo\registry\src\github.com-1ecc6299db9ec823\tonic-build-0.4.2\src\prost.rs:369
>      9: tonic_build::prost::Builder::compile<str>
>                at C:\Users\jeff\.cargo\registry\src\github.com-1ecc6299db9ec823\tonic-build-0.4.2\src\prost.rs:323
>     10: build_script_build::main
>                at .\build.rs:2
>     11: core::ops::function::FnOnce::call_once<fn() -> enum$<core::result::Result<tuple<>, alloc::boxed::Box<Error, alloc::alloc::Global>>, 1, 18446744073709551615, Err>,tuple<>>
>                at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633\library\core\src\ops\function.rs:227
>   note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
> ```

Looks like you are running into https://github.com/tokio-rs/prost/issues/457. I think the issue in our case is the combination of optional fields with an inner `enum`.  We'll probably have to move the enum out of the message for now.

Reply:
Using the oneof suggestion below bypasses this issue.